lib/vmpooler/pool_manager.rb

```diff
@@ -476,29 +550,32 @@ def migrate_vm(vm_name, pool_name, provider)
     end
 
     def _migrate_vm(vm_name, pool_name, provider)
-      $redis.srem('vmpooler__migrating__' + pool_name, vm_name)
+      $redis.srem("vmpooler__migrating__#{pool_name}", vm_name)
 
-      parent_host_name = provider.get_vm_host(pool_name, vm_name)
-      raise('Unable to determine which host the VM is running on') if parent_host_name.nil?
+      provider_name = get_provider_name(pool_name)
+      vm = provider.get_vm_details(pool_name, vm_name)
+      raise('Unable to determine which host the VM is running on') if vm['host'].nil?
       migration_limit = migration_limit $config[:config]['migration_limit']
       migration_count = $redis.scard('vmpooler__migration')
 
-      if !migration_limit
-        $logger.log('s', "[ ] [#{pool_name}] '#{vm_name}' is running on #{parent_host_name}")
-        return
-      elsif migration_count >= migration_limit
-        $logger.log('s', "[ ] [#{pool_name}] '#{vm_name}' is running on #{parent_host_name}. No migration will be evaluated since the migration_limit has been reached")
-        return
-      else
-        $redis.sadd('vmpooler__migration', vm_name)
-        host_name = provider.find_least_used_compatible_host(pool_name, vm_name)
-        if host_name == parent_host_name
-          $logger.log('s', "[ ] [#{pool_name}] No migration required for '#{vm_name}' running on #{parent_host_name}")
+      if migration_limit
+        max_age = 60
+        percentage_of_hosts_below_average = 100
+        run_select_hosts(provider, pool_name, provider_name, vm['cluster'], vm['datacenter'], max_age, percentage_of_hosts_below_average)
+        if migration_count >= migration_limit
+          $logger.log('s', "[ ] [#{pool_name}] '#{vm_name}' is running on #{vm['host']}. No migration will be evaluated since the migration_limit has been reached")
+        elsif $provider_hosts[provider_name][vm['datacenter']][vm['cluster']]['architectures'][vm['architecture']].include?(vm['host'])
+          $logger.log('s', "[ ] [#{pool_name}] No migration required for '#{vm_name}' running on #{vm['host']}")
         else
-          finish = migrate_vm_and_record_timing(vm_name, pool_name, parent_host_name, host_name, provider)
-          $logger.log('s', "[>] [#{pool_name}] '#{vm_name}' migrated from #{parent_host_name} to #{host_name} in #{finish} seconds")
+          $redis.sadd('vmpooler__migration', vm_name)
+          target_host_name = select_next_host(provider_name, vm['datacenter'], vm['cluster'], vm['architecture'])
```

Comment:
what if `select_next_host` returns 'nil'?

Reply:
It should cause a failure to be logged reporting that vm migration failed with an error raised from lib/vmpooler/providers/vsphere.rb:90 when migrate_vm is called. Perhaps it's worth emitting a more clear error that there was no target host provided.

Reply:
I think it will call the next line `migrate_vm_and_record_timing()` with `target_host_name` = nil which seems to go all the way  down to the vcenter API to try and get the host by dns name? 
We should fail earlier than that.

Reply:
This will now fail if the datacenter, or cluster are missing, and if the target is empty, before trying to run the methods with bad data.